src/cephadm/cephadm.py

```diff
@@ -6188,7 +6188,87 @@ def command_deploy(ctx):
     if not ctx.reconfig and not redeploy:
         if ctx.tcp_ports:
             daemon_ports = list(map(int, ctx.tcp_ports.split()))
+    _common_deploy(ctx, daemon_type, daemon_id, daemon_ports, redeploy)
 
+
+def command_deploy_from(ctx: CephadmContext) -> None:
+    """The deploy-from command is similar to deploy but sources nearly all
+    configuration parameters from an input JSON configuration file.
+    """
+    source = '-'
+    if 'source' in ctx and ctx.source:
+        source = ctx.source
+    if source == '-':
+        config_data = json.load(sys.stdin)
+    else:
+        with open(source, 'rb') as fh:
+            config_data = json.load(fh)
+    logger.debug('Loaded deploy configuration: %r', config_data)
+
+    # Bind properties taken from the json to our ctx, similar to how
+    # cli options on `deploy` are bound to the context.
+    ctx.name = config_data['name']
+    for key, value in config_data.get('params', {}).items():
+        setattr(ctx, key, value)
+    if 'image' in config_data:
+        ctx.image = config_data['image']
+    if 'fsid' in config_data:
+        ctx.fsid = config_data['fsid']
+    if 'meta' in config_data:
+        ctx.meta_properties = config_data['meta']
+    if 'config_blobs' in config_data:
+        ctx.config_blobs = config_data['config_blobs']
+        ctx.config_json = config_data['config_blobs']
+    # this is quite the hack!  this is a workaround for the fact that
+    # cephadm orch module knows about "args" not what they represent
+    # and so this lets us (try to) smooth the transition. Ideally, these
+    # would be set using key-value pairs in 'params' above.
+    if 'deploy_arguments' in config_data:
+        parser = argparse.ArgumentParser(add_help=False, exit_on_error=False)
+        _build_parser_args(parser)
+        parsed_args = parser.parse_args(config_data['deploy_arguments'])
+        for key, value in vars(parsed_args).items():
+            setattr(ctx, key, value)
```

Comment:
a bit confused what's going on with this `deploy_arguments`. Originally, it seemed to be added to the `_run_cephadm` call when the call was converted to use `_orch deploy`, then it was made so all the things that were put into it were instead put into `params`, and then when converting to the exchange `Deploy` class it was dropped from the call entirely. Do we actually still need to handle this argument? It seems as if it all goes through `params` in the current state of the PR

Reply:
Sort of. This is one of the bugs that need fixing. This code serves two purposes, one that was intentional and one that was unintentional until I found the bug.  The intentional case is the obvious one, to take command-line argument style values out of the json as if they were passed on the cli, for backwards compatibility. The unintentional but important case is that they also initialize the default values of these attributes of the `ctx`. The bug is that these defaults are set after the loop that assigns `params`. 

We can probably drop `deploy_arguments` as part of the PR, *but* we will need something to replace it for assigning default values. So I wasn't sure if I want to drop it from this PR yet.

Reply:
> being compatible with earlier commits in this PR

Yes

> this is a new field added in one commit in this Pr and then dropped ... in another commit so the end result is it has no effect.

Yes

It ends up serving the need to of initializing some ctx properties and only that. We can make the changes to clean that up in this PR or later, but since this PR doesn't remove the old deploy I felt it simpler to leave it in.  I think it would make more sense to clean it up when the old `deploy` is removed.